lib/src/functions.dart

```diff
@@ -41,6 +41,19 @@ Map<K, V> mergeMaps<K, V>(Map<K, V> map1, Map<K, V> map2,
   return result;
 }
 
+/// Associates the elements in [values] by the value returned by [key].
+///
+/// Returns a map from keys computed by [key] to the last value for which [key]
+/// returns that key.
+Map<T, S> lastBy<S, T>(Iterable<S> values, T Function(S) key) {
```

Comment:
I like having top-level functions, but the reviews I get generally lean towards making functions like this extension methods.

You could put a forwarder into an appropriate existing extension, like [`IterableExtension`](https://github.com/dart-lang/collection/blob/69766daafbaa8535d1343fb7cd87e713f57c107f/lib/src/iterable_extensions.dart#L20), as
```dart
extension IterableExtension<T> on Iterable<T> {
  // ...

  /// Associates the elements in [this] by the value returned by [key].
  ///
  /// Returns a map from keys computed by [key] to the last value for which [key]
  /// returns that key.
  Map<K, T> lastBy<K, T>(K Function(T) key) => functions.lastBy(this, key);

  // ...
}
```

Alternatively, making it only an extension method is also fine (and preferred by some because it doesn't introduce a top-level name that might conflict with something.)

Reply:
I added it as an extension (rather than replacing the top-level version) for consistency with `groupBy` which IMHO is quite similar - happy to move both as extensions if needed.

Reply:
This is fine.
The `groupBy` is covered by the more precise `groupListsBy` and `groupSetsBy` in the extension.